test/unit/master/test_build.py

```diff
@@ -350,13 +350,134 @@ def test_generate_project_type_raises_error_if_failed_to_generate_project(self):
+
+        build.prepare(subjobs=subjobs, job_config=job_config)
+
+        self.assertIsNotNone(build.get_state_timestamp(BuildStatus.PREPARING),
+                             '"preparing" timestamp should not be set before build preparation.')
+        self.assertIsNotNone(build.get_state_timestamp(BuildStatus.PREPARED),
+                             '"prepared" timestamp should not be set before build preparation.')
+
+    def test_allocating_slave_to_build_sets_building_timestamp_only_on_first_slave_allocation(self):
+        mock_slave1 = self._create_mock_slave()
+        mock_slave2 = self._create_mock_slave()
+        build = self._create_test_build(BuildStatus.PREPARED)
+
+        self.assertIsNone(build.get_state_timestamp(BuildStatus.BUILDING),
+                          '"building" timestamp should not be set until slave allocated.')
+
+        build.allocate_slave(slave=mock_slave1)
+        building_timestamp1 = build.get_state_timestamp(BuildStatus.BUILDING)
+        build.allocate_slave(slave=mock_slave2)
+        building_timestamp2 = build.get_state_timestamp(BuildStatus.BUILDING)
+
+        self.assertIsNotNone(building_timestamp1, '"building" timestamp should be set after first slave allocated.')
+        self.assertEqual(building_timestamp1, building_timestamp2,
+                         '"building" timestamp should not change upon further slave allocation.')
+
+    def test_finishing_build_sets_finished_timestamp(self):
+        build = self._create_test_build(BuildStatus.BUILDING)
+
+        self.assertIsNone(build.get_state_timestamp(BuildStatus.FINISHED),
+                          '"finished" timestamp should not be set until build finishes.')
+
+        self._finish_test_build(build)
+        self.assertIsNotNone(build.get_state_timestamp(BuildStatus.FINISHED),
+                             '"finished" timestamp should be set when build finishes.')
+
+    def test_marking_build_failed_sets_error_timestamp(self):
+        build = self._create_test_build(BuildStatus.BUILDING)
+
+        self.assertIsNone(build.get_state_timestamp(BuildStatus.ERROR),
+                          '"error" timestamp should not be set unless build fails.')
+
+        build.mark_failed('Test build was intentionally marked failed.')
+        self.assertIsNotNone(build.get_state_timestamp(BuildStatus.ERROR),
+                             '"error" timestamp should be set when build fails.')
+
+    def test_canceling_build_sets_canceled_timestamp(self):
+        build = self._create_test_build(BuildStatus.BUILDING)
+
+        self.assertIsNone(build.get_state_timestamp(BuildStatus.CANCELED),
+                          '"canceled" timestamp should not be set unless build is canceled.')
+
+        build.cancel()
+        self.assertIsNotNone(build.get_state_timestamp(BuildStatus.CANCELED),
+                             '"canceled" timestamp should be set when build is canceled.')
+
+    # todo: almost all tests in this file should use this helper.
```

Comment:
I'll refactor the rest of the tests in this file to use the added helper code in a separate commit.